libsigner/src/events.rs

```diff
@@ -61,14 +63,106 @@ const SIGNATURE_SHARE_REQUEST_SLOT_ID: u32 = 8;
 /// The slot ID for the block response for miners to observe
 pub const BLOCK_SLOT_ID: u32 = 10;
+/// The slot ID for NACK messages to signers with outdated view of the chain
+pub const NACK_SLOT_ID: u32 = 11;
 
 /// The messages being sent through the stacker db contracts
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
 pub enum SignerMessage {
     /// The signed/validated Nakamoto block for miners to observe
     BlockResponse(BlockResponse),
     /// DKG and Signing round data for other signers to observe
-    Packet(Packet),
+    Packet(PacketInfo),
+    /// NACK messages to signers with outdated view of the chain
+    Nack(NackMessage),
+}
+
+/// Signed network packets sent by signers along with their view
+/// of stacks tip consensus hash and block height for
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct PacketInfo {
+    /// Signed network packet
+    pub packet: Packet,
+    /// The signer's view of stacks tip consensus hash
+    pub consensus_hash: String,
+    /// The signer's view of stacks tip block height
+    pub block_height: u64,
+}
+
+impl PacketInfo {
+    /// Create a new PacketInfo
+    pub fn new(packet: Packet, consensus_hash: String, block_height: u64) -> Self {
+        PacketInfo {
+            packet,
+            consensus_hash,
+            block_height,
+        }
+    }
+}
+
+/// NACK message structure
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct NackMessage {
+    /// The reporting signer_id having a more updated chain view
+    pub signer_id: u32,
+    /// The signer's view of calculated coordinator based on its view of the chain
+    pub coordinator_info: CoordinatorInfo,
+}
+
+impl NackMessage {
+    /// Creates a new `NackMessage` with the given coordinator information
+    pub fn new(signer_id: u32, coordinator_info: CoordinatorInfo) -> Self {
+        Self {
+            signer_id,
+            coordinator_info,
+        }
+    }
+}
+
+/// Calculated Coordinator Metadata per signer's view of Stacks chain
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CoordinatorInfo {
+    /// The calculated coordinator's id - same as signer-id
+    pub coordinator_id: u32,
+    /// The calculated coordinator's public key
+    pub coordinator_public_key: PublicKey,
+    /// The stacks_tip_consensus_hash from the signer's view
+    pub stacks_consensus_hash: String,
+    /// The stacks tip block height from the signer's view
+    pub stacks_block_height: u64,
+}
+
+impl CoordinatorInfo {
+    /// Create a new `CoordinatorInfo` instance
+    pub fn new(
+        coordinator_id: u32,
+        coordinator_public_key: PublicKey,
+        stacks_consensus_hash: String,
+        stacks_block_height: u64,
+    ) -> Self {
+        Self {
+            coordinator_id,
+            coordinator_public_key,
+            stacks_consensus_hash,
+            stacks_block_height,
+        }
+    }
+}
+
+impl PartialEq for CoordinatorInfo {
```

Comment:
I wonder can this be auto inferred or does PublicKey not use PartialEq?

Reply:
yep, PublicKey doesn't

Reply:
Yeah I ran into this in my PR.  As a workaround I passed around Points instead of PublicKeys.